tests/kernel/usleep/src/main.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2019 Intel Corp.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <zephyr.h>
+
+/*
+ * precision timing tests in an emulation environment are not reliable.
+ * if the test passes at least once, we know it works properly, so we
+ * attempt to repeat the test RETRIES times before reporting failure.
+ */
+
+#define RETRIES		10
+
+/*
+ * Theory of operation: we can't use absolute units (e.g., "sleep for 10us")
+ * in testing k_usleep() because the granularity of sleeps is highly dependent
+ * on the hardware's capabilities and kernel configuration. Instead, we
+ * test that k_usleep() actually sleeps for the minimum possible duration.
+ * (That minimum duration is presently two ticks; see below.) So, we loop
+ * k_usleep()ing for as many iterations as should comprise a second, and
+ * check to see that a total of one second has elapsed.
+ */
+
+#define LOWER_BOUND_MS	900		/* +/- 10%, might be too lax */
+#define UPPER_BOUND_MS	1100
+
+void test_usleep(void)
+{
+	int retries = 0;
+	s64_t elapsed_ms;
+
+	while (retries < RETRIES) {
+		s64_t start_ms;
+		s64_t end_ms;
+		int i;
+
+		++retries;
+		start_ms = k_uptime_get();
+
+		for (i = 0; i < (CONFIG_SYS_CLOCK_TICKS_PER_SEC / 2); ++i) {
+			/*
+			 * this will always sleep for TWO ticks:
+			 *
+			 * the conversion from 1us to ticks is rounded
+			 * up to the nearest tick boundary, and sleeps
+			 * always have _TICK_ALIGN (currently 1) added
+			 * to their durations.
```

Comment:
Hm.... that shouldn't be correct.  The _TICK_ALIGN nonsense is there to make sure we always suspend for at least the requested delay.  In effect it's there to correct a off-by-one error in the way we originally specified the clock driver timeout argument (and which I didn't try to fix during the rework because the timing internals were originally run against the older drivers before the drivers got rewritten).  But in the case where our delays are in microseconds, it shouldn't be doubling the (much larger) tick count.  The intent of the API is that you get the "next tick boundary after the requested delay".  If that's not what's actually happening, this sounds like a bug.

Reply:
Well, the comment is perhaps a bit misleading.  It's a worst of worst cases.

`_TICK_ALIGN` is added because we have no idea how soon in the future the next tick will arrive, so we assume the next tick is imminent, i.e, a sleep of 1 tick will sleep for almost 0 time.  This technique adds in an uncertainty of 0-1 ticks to the intended sleep duration, on average, 0.5 ticks.  The worst case of 1 tick occurs when we sleep in synchronization with the system clock, and that's precisely what happens in this test case: we're going to sleep immediately after the last tick woke us up (guaranteed on every iteration but the first).

Since we have to assume that `_TICK_ALIGN` adds a delay of 0, then we have to round up the provided sleep time to at least a tick boundary, lest we return from the sleep earl.  Hence the use of `ceiling_fraction()` in the division.  The smaller the requested sleep time is relative to the tick duration, the more pronounced the error introduced by this rounding.  In the test case, `k_usleep()` is purposely executing a sleep at the limits of the hardware resolution, showing the most pronounced distortion of the duration.

So these two factors in combination make this test case a really poor illustration of the actual behavior of `k_usleep()` et al.  When we're not synchronized to the system clock and using sleep times that are not near the limits of the hardware clock, the behavior is much more reasonable. For example, with a tick duration of 10us and a requested sleep of 100us, we should see an average sleep time of 105us.  Not so bad.

There are some other solutions.  For example, we could inspect the timer hardware instead of using a worst-case `_TICK_ALIGN`: this is trickier than it might seem at first because of possible races.  Plus there may be some platforms on which such inspection isn't possible.  Alternatively, we could change the API the allow sleeps to return early.  All in all, though, I think the easiest fix is to simply inform users that they should use  a `CONFIG_SYS_CLOCK_TICKS_PER_SEC` that is >> their desired uncertainty.  This basically comes at no cost in a `TICKLESS` kernel.